src_lockfree/htbl/htbl.body.ml

```diff
@@ -0,0 +1,617 @@
+            rest = Nil;
+          }
+      in
+      if Atomic_array.unsafe_compare_and_set r.buckets i (B before) (B after)
+      then (
+        Size.update_once r.size cons_r.size_modifier;
+        cons_r.size_modifier <- Size.used_once;
+        adjust_size t r mask true)
+      else try_add t key value (Backoff.once backoff)
+  | B (Cons_with_size cons_r as before) ->
+      if cons_r.size_modifier != Size.used_once then
+        Size.update_once r.size cons_r.size_modifier;
+      if r.equal cons_r.key key then false
+      else if assoc_node r.equal key cons_r.rest != Nil then false
+      else
+        let (Cons_with_size cons_r as after) : ('a, 'b, [ `Cons_with_size ]) tdt
+            =
+          Cons_with_size
+            {
+              size_modifier = Size.(new_once r.size incr);
+              key;
+              value;
+              rest =
+                Cons
+                  { key = cons_r.key; value = cons_r.value; rest = cons_r.rest };
+            }
+        in
+        if Atomic_array.unsafe_compare_and_set r.buckets i (B before) (B after)
+        then (
+          Size.update_once r.size cons_r.size_modifier;
+          cons_r.size_modifier <- Size.used_once;
+          adjust_size t r mask true)
+        else try_add t key value (Backoff.once backoff)
+  | B (Resize _) ->
+      let _ = finish t (Atomic.get t) in
+      try_add t key value Backoff.default
+
+let[@inline] try_add t key value = try_add t key value Backoff.default
+
+(* *)
+
+let[@tail_mod_cons] rec dissoc t key = function
+  | Nil -> raise_notrace Not_found
+  | Cons r ->
+      if t key r.key then r.rest else Cons { r with rest = dissoc t key r.rest }
+
+let rec remove_node t key backoff =
+  let r = Atomic.get t in
+  let h = r.hash key in
+  let mask = Atomic_array.length r.buckets - 1 in
+  let i = h land mask in
+  match Atomic_array.unsafe_fenceless_get r.buckets i with
+  | B (Nil_with_size nil_r) ->
+      if nil_r.size_modifier != Size.used_once then (
+        Size.update_once r.size nil_r.size_modifier;
+        nil_r.size_modifier <- Size.used_once);
+      Nil
+  | B (Cons_with_size cons_r as before) -> begin
+      if cons_r.size_modifier != Size.used_once then
+        Size.update_once r.size cons_r.size_modifier;
+
+      let size_modifier = Size.new_once r.size Size.decr in
+      if r.equal cons_r.key key then
+        let found_node =
+          Cons { key = cons_r.key; value = cons_r.value; rest = Nil }
+        in
+        match cons_r.rest with
+        | Nil ->
+            let (Nil_with_size nil_r as after) :
+                ('a, 'b, [ `Nil_with_size ]) tdt =
+              Nil_with_size { size_modifier }
+            in
+            if
+              Atomic_array.unsafe_compare_and_set r.buckets i (B before)
+                (B after)
+            then (
+              Size.update_once r.size nil_r.size_modifier;
+              nil_r.size_modifier <- Size.used_once;
+              adjust_size t r mask found_node)
```

Comment:
These three lines of code could probably be nicely folded into a single line:
```ocaml
adjust_size t r after mask found_node
```
The `adjust_size` function would then perform the `Size.update` on `after`.

Note that the native OCaml compiler can optimize the access of `size_modifier` such that no case discrimination (between `Nil_with_size` and `Cons_with_size`) is required as long as the field is at the same offset and there no other constructors are allowed.  Here is [an example on godbolt](https://ocaml.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:12,fontUsePx:'0',j:1,lang:ocaml,selection:(endColumn:1,endLineNumber:12,positionColumn:1,positionLineNumber:12,selectionStartColumn:1,selectionStartLineNumber:12,startColumn:1,startLineNumber:12),source:'type+_+t+%3D%0A++%7C+Foo+:+%7B+mutable+x+:+int%3B+foo+:+bool+%7D+-%3E+%5B%3E+%60Foo%5D+t%0A++%7C+Bar+:+%7B+mutable+x+:+int%3B+bar+:+int+list+%7D+-%3E+%5B%3E+%60Bar%5D+t%0A++%7C+Extra_ball+:+%5B%3E+%60Extra_ball%5D+t%0A%0A(*+The+write+here+gets+translated+to+a+single+mov+--+no+pattern+matching+*)%0Alet+set+n+:+%5B%3C+%60Foo+%7C+%60Bar%5D+t+-%3E+_+%3D+function%0A++%7C+Foo+r+-%3E+r.x+%3C-+n%0A++%7C+Bar+r+-%3E+r.x+%3C-+n%0A%0A(*+The+read+here+gets+translated+to+a+single+mov+--+no+pattern+matching+*)%0Alet+get+:+%5B%3C+%60Foo+%7C+%60Bar%5D+t+-%3E+_+%3D+function%0A++%7C+Foo+r+-%3E+r.x+%0A++%7C+Bar+r+-%3E+r.x%0A'),l:'5',n:'1',o:'OCaml+source+%231',t:'0')),k:51.05977810895844,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:ocaml5200,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:ocaml,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+ocamlopt+5.2.0+(Editor+%231)',t:'0')),header:(),k:48.940221891041574,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) that shows that the native compiler indeed optimizes such code.

Reply:
I am not sure to get it. Is this just factorizing code by adding the size update on after in  `adjust_size` code  ?

Reply:
Yes, the idea here is to reduce code size as the two lines (i.e. `Size.update_once ...; r.size_modifier <- ...` could be moved to happen in `adjust_size`.  If you look through the code, you will notice that every places that calls `adjust_size` also has such two lines.

The note with link to godbolt is to show that OCaml specifically does an optimization that makes this cheap.  The issue being that in some cases you have a `Nil_with_size` and in other case sa `Cons_with_size`.  Thanks to the optimization (and use of GADTs), you can write the code in such a way that the native OCaml compiler can completely avoid the pattern matching to distinguish between those two cases as the `size_modifier` field is at the same relative position in both cases.